mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoQueryProviderMethodGenerator.java

```diff
@@ -0,0 +1,191 @@
+          FUTURE_OF_ENTITY,
+          FUTURE_OF_OPTIONAL_ENTITY,
+          RESULT_SET,
+          FUTURE_OF_ASYNC_RESULT_SET,
+          PAGING_ITERABLE,
+          FUTURE_OF_ASYNC_PAGING_ITERABLE);
+
+  public DaoQueryProviderMethodGenerator(
+      ExecutableElement methodElement,
+      DaoImplementationSharedCode enclosingClass,
+      ProcessorContext context) {
+    super(methodElement, enclosingClass, context);
+  }
+
+  @Override
+  public Optional<MethodSpec> generate() {
+
+    // Validate the return type:
+    ReturnType returnType = parseReturnType(methodElement.getReturnType());
+    if (!SUPPORTED_RETURN_TYPES.contains(returnType.kind)) {
+      context
+          .getMessager()
+          .error(
+              methodElement,
+              "Invalid return type: %s methods must return void, boolean, Integer, Row, "
+                  + "an entity class, a result set, a mapped iterable, or a "
+                  + "CompletionStage/CompletableFuture of any of the above",
+              QueryProvider.class.getSimpleName());
+      return Optional.empty();
+    }
+
+    // Generate the method.
+    MethodSpec.Builder methodBuilder = GeneratedCodePatterns.override(methodElement);
+
+    // - request the instantiation of the provider during DAO initialization
+    List<ClassName> entityHelperTypes = getEntityHelperTypes();
+    String providerName =
+        enclosingClass.addQueryProvider(methodElement, getProviderClass(), entityHelperTypes);
+
+    // - delegate to the provider's method
+    String providerMethod = methodElement.getAnnotation(QueryProvider.class).providerMethod();
+    if (providerMethod.isEmpty()) {
+      providerMethod = methodElement.getSimpleName().toString();
+    }
+
+    methodBuilder.addCode("$[");
+    if (returnType.kind != VOID) {
+      methodBuilder.addCode("return ");
+    }
+    methodBuilder.addCode("$L.$L(", providerName, providerMethod);
+    boolean first = true;
+    for (VariableElement parameter : methodElement.getParameters()) {
+      if (first) {
+        first = false;
+      } else {
+        methodBuilder.addCode(", ");
+      }
+      methodBuilder.addCode("$L", parameter.getSimpleName().toString());
+    }
+    methodBuilder.addCode(");$]\n");
+
+    return Optional.of(methodBuilder.build());
+  }
+
+  private TypeMirror getProviderClass() {
+    AnnotationMirror annotationMirror = getQueryProviderAnnotationMirror();
+    for (Map.Entry<? extends ExecutableElement, ? extends AnnotationValue> entry :
+        annotationMirror.getElementValues().entrySet()) {
+      if (entry.getKey().getSimpleName().contentEquals("providerClass")) {
+        return ((TypeMirror) entry.getValue().getValue());
+      }
+    }
+    // providerClass is mandatory on the annotation, so if we get here the user will already have to
+    // deal with a compile error.
+    // But return something so that the processor doesn't crash downstream.
+    return context.getTypeUtils().getPrimitiveType(TypeKind.INT);
```

Comment:
Why we are choosing `INT`? Wouldn't it confuse users?

Reply:
See the comment.